### PR TITLE
VCC-3958 Allow `disabled` on `<select>` or in Combobo config

### PR DESCRIPTION
**Pull request by stephymiehle**
This allows the entire Combobo to be disabled. If `disabled` is present on a `<select>`, that will be used; otherwise, we'll check the config.

Undoing the interactivity was more complex than expected. Select-only Comobobo use `<div>`, which cannot support `disabled` (but can support `aria-disabled`), and interactivity needs to be prevented via keyboard and mouse too.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -816,6 +964,187 @@ module.exports = class Combobo {
...
 
+  /**
+   * Clear all options from the Combobo. This will remove all options from the Combobo and
+   * reset the Combobo to its initial state.
+   * 
+   * @param {Object} args
+   * @param {String} [args.placeholder] The placeholder text to be added back to the Combobo.
+   * @param {Boolean} [args.usePlaceholder] If true, the placeholder option that's added back will be selected.
+   */
+  clearOptions(args = {}) {
```
Comment (Farsad) [inline index.js:985]: @stephymiehle  I noticed the addOptions and clearOptions methods in the recent changes. I wasn’t able to find their usages in the current implementation. If these methods were intended as improvements, would it be possible to include them in a separate PR and commit for better clarity?

Additionally, are the other functionalities tested when options are added or removed dynamically? If addOptions and clearOptions are not intended for internal use, perhaps we could consider updating the documentation to reflect their purpose.

**Review by Farsad — COMMENTED** [review]
Hey @stephymiehle.

Have you had a chance to run `yarn test:dev` on this branch yet? I’m a bit concerned that these extensive changes might affect the tests.

Additionally, it would be really helpful to see more implementations of the <select> element in the demo, rather than just the HTML structures. This way, users could have a better reference point.
Thanks!

**Comment by stephymiehle** [comment]
Thanks @Farsad! Yes, all tests are passing. I separated the new helper methods into a separate commit (with a force push so the line comments are better). They were added while working on the LMI stuff, but I agree that they make sense to mention separately.

I have a separate update for the demo page that I'll make as a separate PR; it demonstrates how to add options dynamically.

---

```
y test:dev                                         
yarn run v1.22.19
$ mochify --recursive --watch --transform [ babelify --presets [ es2015 ] ] --transform [ stringify --extensions [ .html .css ] ] --plugin [ proxyquire-universal ]
# chromium:


  Combobo
    config
      ✅ should call lib/config
      given truthy useLiveRegion
        ✅ should create the assertive live region
      groups
        ✅ should properly configure the groups property
    error handling
      ✅ should throw if input/list are unable to be found
    attrs
      ✅ should call lib/attributes
    initial state
      ✅ should properly set initial state
    events
      input clicks
        ✅ should open the list (581ms)
      input blur
        given no hover
          ✅ should close the list
      input focus
        given at least 1 selection
          ✅ should properly set the input's value
      document clicks
        given a target outside of Combobo
          ✅ should close the list
      optionEvents
        option clicks
          ✅ should goTo/select the clicked option
        option mouseovers
          ✅ should properly handle activeClass
          ✅ should set isHovering to true
        option mouseouts
          ✅ should properly handle activeClass
          ✅ should set isHovering to false
      keyboard events
        input keydowns
          UP/DOWN
            given a closed list
              ✅ should open the list and go to the first option
            given a open list
              UP
                ✅ should go to the previous option
              DOWN
                ✅ should go to the next option
          ENTER
            ✅ should select the currently active option
          ESCAPE
            ✅ should close the list
          BACKSPACE
            given multiple options selected
              ✅ should clear the input's value
        input keyups
          given a truthy freshSelection
            ✅ should reset everything
            given an altered input value
              ✅ should filter the options, open the list, and set freshSelection to false
          given a falsey freshSelection
            ✅ should open ensure the list is opened/filtered
          no results
            given a config with noResults set
              ✅ should call no lib/no-results
    openList
      ✅ should be chainable
      ✅ should add the openClass
      ✅ should set aria-expanded to "true"
      ✅ should announce the count if the list was not open
      ✅ should set isOpen to true
      ✅ should emit the "list:open" event
      given a list outside the viewport
        ✅ should call scrollToElement
    closeList
      ✅ should be chainable
      ✅ should remove the openClass
      ✅ should set aria-expanded to "false"
      ✅ should emit the "list:close" event
      given a truthy "focus" param
        ✅ should focus the input
      given falsey multiselect and a selection
        ✅ should properly set the input's value
    select
      ✅ should be chainable
      ✅ should set the aria-activedescendant attribute properly
      ✅ should emit the selection event
      ✅ should properly set aria-selected
      given single select
        ✅ should remove the selectedClass from the previously selected option and add it to the new one
        ✅ should clobber previous selections and set this.selected properly
        ✅ should close the list when a selection is made
        given a truthy `allowEmpty` option
          ✅ should clear the `selected` array
      given mulitselect
        ✅ should toggle selectedClass properly
        ✅ should properly emit the deselection event
        ✅ should take a deselection out of the array
        ✅ should NOT close the list when a selection is made
    clearFilters
      ✅ should unhide all options/groups
      ✅ should update currentOpts
    reset
      ✅ should clear the value of the input
      ✅ should remove aria-activedescendant from the input
      ✅ should remove data-active-option from the input
      ✅ should remove active style from all options
      ✅ should set currentOption to null
      ✅ should set selected to an empty array
      ✅ should remove the selectedClass from all options
      ✅ should remove the activeClass from all options
      ✅ should set aria-selected to false on all options
    filter
      given a funktion
        ✅ should properly filter
      given a string (preset)
        ✅ should call the right preset
      given a change in options
        ✅ should announce the count
    announceCount
      given a truthy useLiveRegion
        ✅ should call liveRegion.announce
    emptyDropdownList
      ✅ should clear the dropdown list options
    setCurrentOption
      ✅ should set the current option
    setOptions
      ✅ should add an option to the dropdown list
    updateSelectedOptions
      ✅ should update selected Options if any

  lib/announce-active
    ✅ should be a function
    given a falsey groupChanged
      ✅ should invoke announce with the proper arguments
    given a truthy groupChanged, config announcement groupChange
      ✅ should properly prepend the groupChange text to the announcement
    given a truthy selected
      ✅ should append the proper selected text to the message

  lib/attributes
    ✅ should be a function
    ✅ should add the right attributes

  lib/config
    ✅ should be a function
    ✅ should properly merge the userConfig with the defaults

  lib/current-group
    ✅ should be a function
    ✅ should return the proper group

  lib/filter
    contains
      ✅ should properly filter the options
    equals
      ✅ should properly filter the options
    starts-with
      ✅ should properly filter the options

  Initializing from select with multiple attribute
    ✅ should assert the options in the list
    ✅ should select the first option from the list
    ✅ should select the second option from the list

  lib/no-results
    given noResultsText, no currentOpts and no no-results element
      ✅ should create the proper element and append it to the list
    given current options and an existing noResults element
      ✅ should remove the noResults element

  lib/utils/rndid
    ✅ should generate a random 8-character id
    ✅ should not be null
    ✅ should be unique
    ✅ should start with a letter

  lib/utils/wrap-match
    ✅ should properly wrap the match

  lib/value
    ✅ should support data-value attribute (as the 1st priority)
    ✅ should support innerText (as the 2nd priority)


  94 passing (1s)
```

**Comment by Farsad** [comment]
@stephymiehle I see this error:

Screenshot 2024-06-10 at 15 48 32

**Comment by stephymiehle** [comment]
@Farsad Hmm, tests are still passing for me locally. I wonder if we have a discrepancy somewhere? I just updated the yarn lockfile in case we have different versions. Could you please try again?

**Review by Farsad — APPROVED** [review]

**Review by vicaraya — APPROVED** [review]

**Pull request merged**